kratos/python/add_variable_utils_to_python.cpp

```diff
@@ -452,6 +452,63 @@ void AddVariableUtilsToPython(pybind11::module &m)
         .def("ApplyFixity", ApplyFixity<Variable<double>>)
         .def("ApplyFixity", ApplyFlaggedFixity<Variable<double>>)
         .def("ApplyVector", &VariableUtils::ApplyVector<Variable<double>>)
+        .def("ApplyVector", []( VariableUtils& rVariableUtils, 
+                                const Variable<array_1d<double,3>>& rVar,
+                                unsigned int Dim,
+                                const Vector& rData,
+                                ModelPart::NodesContainerType& rNodes
+                                )
+                {
+                    if(rNodes.size() != 0 && rNodes.size()*Dim == rData.size()) {
+                        // First we do a check
+                        rVariableUtils.CheckVariableExists(rVar, rNodes);
+
+                        IndexPartition<std::size_t>(rNodes.size()).for_each([&](std::size_t index){
+                            auto it_node = rNodes.begin() + index;
+                            auto& v = it_node->FastGetSolutionStepValue(rVar);
+                            for(unsigned int k=0; k<Dim; k++)
+                                v[k] = rData[index*Dim+k];
+                        });
+                    } else
+                        KRATOS_ERROR  << "There is a mismatch between the size of data array and the number of nodes ";
+                })
+        .def("GetSolutionStepValuesVector", []( VariableUtils& rVariableUtils, 
+                                const Variable<array_1d<double,3>>& rVar,
+                                unsigned int Dimension,
```

Comment:
For the sake of consistency, I'd remove the Dimension and always use 3. See `Node<3>`, `SetVariableToZero`, `array_1d<double,3>`, etc.

At least, I'd move `Dimension` to the last position and set `py::arg("Dimension") = 3`.

Reply:
no, i explicitly do not want to do this. the point is that i want to be able to extract only the x y components if i am in 2D.

this is really useful because  this way one can write vectorized operations without having to operate on the third component. 

for example this allows me to write a tool with multiple relative rotations around the axis Z (i need it for a project but it is quite specific as)

~~~py
# Importing the Kratos Library
import KratosMultiphysics
import math
import numpy as np

def Factory(settings, Model):
    if not isinstance(settings, KratosMultiphysics.Parameters):
        raise Exception("expected input shall be a Parameters object, encapsulating a json string")
    return RotationHelperProcess(Model, settings["Parameters"])

class RotationHelperProcess(KratosMultiphysics.Process):
    """This process applies a rotation around the axis Z to all of the subparts involved in the project.
    Then each of the subparts is also allowed to have a rotation around its own axis (and possibly at different roational speed)
    """

    def ComputeRotationMatrix(self,theta):
        s = np.sin(theta)
        c = np.cos(theta)
        R = np.array([[c, -s],[s,c]])
        return R

    def ApplyRotation(self,xc0, R, nodes, disp_var):

        var_utils = KratosMultiphysics.VariableUtils();
        #[n1.x0 n1.y0  n2.x0 n2.y0  n3.x0 n3.y0 ....]
        dx = np.array(var_utils.GetInitialPositionsVector(nodes, 2)) #, copy=False)

        #[[n1.x0 n1.y0  
        #  n2.x0 n2.y0
        #  n3.x0 n3.y0 
        #  ....]
        dx = dx.reshape(len(nodes),2,order="c")
        
        #[[n1.x0-xc[0] n1.y0-xc[1]  
        #  n2.x0-xc[0] n2.y0-xc[1]
        #  n3.x0-xc[0] n3.y0 -xc[1]
        #  ....]
        dx -= xc0

        #x-xc0 = R@(x0-xc0) -> x = xc0 + R@(x0-xc0) -> u = x - x0 = -(x0-xc0) + R@(x0-xc0)
        #(R*positions0.T).T = positions0@R.T
        u = dx@R.T 
        u -= dx
        u = KratosMultiphysics.Vector(u.reshape(len(nodes)*2))

        var_utils.ApplyVector(disp_var, 2, u, nodes)
        # for node in nodes:
        #     dx = np.array([node.X0-xc0[0], node.Y0-xc0[1]])

        #     #x-xc0 = R@(x0-xc0) -> x = xc0 + R@(x0-xc0) -> u = x - x0 = -(x0-xc0) + R@(x0-xc0)
        #     u = -dx 
        #     u += R@dx
        #     aux = np.array([u[0],u[1],0.0])
        #     node.SetSolutionStepValue(disp_var,0,aux)

    def GetDefaultSettings(self):
        return KratosMultiphysics.Parameters("""
                {
                    "displacement_variable"            : "MESH_DISPLACEMENT",
                    "velocity_variable"                : "MESH_VELOCITY",
                    "initial_xy_coords_global_rotation_center" : [0.0, 0.0],
                    "base_model_part_name"             : "AAA",
                    "rpms" : 10.0, //roations/minute
                    "parts"       : [
                        { "part_name" : "name_of_part",
                        "initial_xy_coords_part_center_of_rotation" : [0.0,0.0],
                        "rpms" : 10.0
                        }
                    ]
                }""")




    def __init__(self, Model, settings ):
        """ The default constructor of the class

        Keyword arguments:
        self -- It signifies an instance of a class.
        Model -- the container of the different model parts.
        settings -- Kratos parameters containing solver settings.
        """
        KratosMultiphysics.Process.__init__(self)

        self.var_utils = KratosMultiphysics.VariableUtils()

        #The value can be a double or a string (function)
        default_settings = self.GetDefaultSettings()

        settings.ValidateAndAssignDefaults(default_settings)

        #global data
        self.model_part = Model[settings["base_model_part_name"].GetString()]
        self.disp_var = KratosMultiphysics.KratosGlobals.GetVariable(settings["displacement_variable"].GetString())
        self.vel_var = KratosMultiphysics.KratosGlobals.GetVariable(settings["velocity_variable"].GetString())
        self.global_angular_velocity = settings["rpms"].GetDouble()*2*math.pi/60
        self.xc0 = settings["initial_xy_coords_global_rotation_center"].GetVector()

        #data for each blade
        self.part_data = []
        for part in settings["parts"]:
            mp = Model[part["part_name"].GetString()]
            xc0 = np.array( part["initial_xy_coords_part_center_of_rotation"].GetVector() )
            w = part["rpms"].GetDouble()*2*math.pi/60
            self.part_data.append([mp,xc0,w])

        #obtain the list of all nodes in parts involved in the rotation
        #and save the initial coordinates list
        #self.initial_coords_list = []
        self.all_nodes = KratosMultiphysics.NodesArray() 
        for part in self.part_data:
            for node in part[0].Nodes:
                self.all_nodes.append(node)
                #self.initial_coords_list.append(np.array([node.X0,node.Y0,node.Z0]))
        self.initial_coords_list = self.var_utils.GetInitialPositionsVector(self.all_nodes,2)

        #measure the min and max
        self.xmin=node.X0
        self.ymin=node.Y0
        self.zmin=node.Z0
        self.xmax=self.xmin
        self.ymax=self.ymin
        self.zmax=self.zmin
        for node in self.all_nodes:
            self.xmin=min(self.xmin,node.X0)
            self.ymin=min(self.ymin,node.Y0)
            self.zmin=min(self.zmin,node.Z0)
            self.xmax=max(self.xmax,node.X0)
            self.ymax=max(self.ymax,node.Y0)
            self.zmax=max(self.zmax,node.Z0)
        sumx=abs(self.xmin)+abs(self.xmax)
        sumy=abs(self.ymin)+abs(self.ymax)
        sumz=abs(self.zmin)+abs(self.zmax)
        print("x limits",self.xmin,self.xmax, " sum", sumx, " xc = ", self.xmin+sumx/2, " diam=",sumx/2.0)
        print("y limits",self.ymin,self.ymax, " sum", abs(self.ymin)+abs(self.ymax))
        print("z limits",self.zmin,self.zmax, " sum", abs(self.zmin)+abs(self.zmax))

    def ResetInitialPositions(self):
        self.var_utils.SetInitialPositionsVector(self.all_nodes,self.initial_coords_list)
        # for node,initial_pos in zip(self.all_nodes, self.initial_coords_list):
        #     node.X0 = initial_pos[0]
        #     node.Y0 = initial_pos[1]
        #     node.Z0 = initial_pos[2]

    def ComputeTotalDisplacementAndResetInitialPositions(self):
        d = self.var_utils.GetCurrentPositionsVector(self.all_nodes,2) - self.initial_coords_list
        self.var_utils.SetSolutionStepValuesVector(self.disp_var,0,d,self.all_nodes)

        self.var_utils.SetInitialPositionsVector(self.all_nodes,self.initial_coords_list)

        # d = KratosMultiphysics.Array3()
        # for node,initial_pos in zip(self.all_nodes, self.initial_coords_list):
        #     node.X0 = initial_pos[0]
        #     node.Y0 = initial_pos[1]
        #     node.Z0 = initial_pos[2]
        #     d[0] = node.X - node.X0
        #     d[1] = node.Y - node.Y0
        #     d[2] = node.Z - node.Z0

        #     node.SetSolutionStepValue(self.disp_var, 0, d)


    def ExecuteInitializeSolutionStep(self):
        
        """This method is executed in order to initialize the current step

        Keyword arguments:
        self -- It signifies an instance of a class.
        """
        current_time = self.model_part.ProcessInfo[KratosMultiphysics.TIME]
        dt = self.model_part.ProcessInfo[KratosMultiphysics.DELTA_TIME]
        #self.ResetInitialPositions()
        print("current_time",current_time)
        #apply rotation to all of the blades which compose the system
        #note that we TEMPORARILY actualize the InitialCoordinates
        for part in self.part_data:
            theta_part = current_time*part[2]
            xc0_part = part[1]
            nodes = part[0].Nodes
            Rpart = self.ComputeRotationMatrix(theta_part)
            self.ApplyRotation(xc0_part, Rpart, nodes, self.disp_var)
            x0vec = self.var_utils.GetInitialPositionsVector(nodes,2)
            dvec = self.var_utils.GetSolutionStepValuesVector(self.disp_var,2,0,nodes)
            x0vec += dvec
            self.var_utils.SetInitialPositionsVector(nodes, x0vec)
            # for node in nodes: #actualize the initial position to later apply the global rotation
            #     d = node.GetSolutionStepValue(self.disp_var)
            #     node.X0 += d[0]
            #     node.Y0 += d[1]
            #     node.Z0 += d[2]
            
        theta = current_time*self.global_angular_velocity
        R = self.ComputeRotationMatrix(theta) #global rotation matrix
        self.ApplyRotation(self.xc0, R, self.all_nodes, self.disp_var)
        x0vec = self.var_utils.GetInitialPositionsVector(self.all_nodes,2)
        dvec = self.var_utils.GetSolutionStepValuesVector(self.disp_var,2,0,self.all_nodes)
        x0vec += dvec
        self.var_utils.SetCurrentPositionsVector(self.all_nodes, x0vec)
        # for node in self.all_nodes: #actualize the initial position to later apply the global rotation
        #     d = node.GetSolutionStepValue(self.disp_var,0)
        #     node.X = node.X0 + d[0]
        #     node.Y = node.Y0 + d[1]
        #     node.Z = node.Z0 + d[2]

        self.ComputeTotalDisplacementAndResetInitialPositions()

        v = self.var_utils.GetSolutionStepValuesVector(self.disp_var,2,0,self.all_nodes)
        v -= self.var_utils.GetSolutionStepValuesVector(self.disp_var,2,1,self.all_nodes)
        v /= dt
        self.var_utils.SetSolutionStepValuesVector(self.vel_var,0,v,self.all_nodes)
        # for node in self.all_nodes: 
        #     #compute velocity
        #     v = d - node.GetSolutionStepValue(self.disp_var,1)
        #     v/=dt 
        #     node.SetSolutionStepValue(self.vel_var,0,v)

        self.ResetInitialPositions()

~~~

which is hundrueds of time faster than doing it with the loops in python